Avoid re-creating the random number generator in RandomSampler

lartpang opened the pull request:
To optimize the construction of iterator and make behavior consistent with generator or without it, I change `self.generator = generator` when `self.generator is None`.
So, when new iter is called, sampler won't need to re-create a new generator for us.

As mentioned here: https://github.com/pytorch/pytorch/pull/52956#issuecomment-788201006

lartpang committed:
Fixed the error of generator in the RandomSampler.

In  `__iter__` of the `RandomSampler`, when `self.replacement` is `False` in the original code, `self.generator` is always used in the `torch.randperm` instead of the generator we set.

lartpang committed:
Merge pull request #1 from pytorch/master

update to the newest state

lartpang committed:
Update the __iter__ of the RandomSampler

To optimize the construction of iterator and make behavior consistent with generator or without it, I change `self.generator = generator` when `self.generator is None`.
So, when new iter is called, sampler won't need to re-create a new generator for us.

ejguan commented on torch/utils/data/sampler.py:
```diff
@@ -114,6 +114,7 @@ def __iter__(self):
             generator = torch.Generator()
             generator.manual_seed(int(torch.empty((), dtype=torch.int64).random_().item()))
+            self.generator = generator
         else:
             generator = self.generator
```
This part can be eliminated. And the following generator can be also replaced by self.generator.

lartpang replied:
You are right.

In addition, would it be more appropriate if we moved the following `if` code into `__init__` of `RandomSampler`?  That is, `self.generator` is instantiated only when the class `RandomSampler` is initialized.

i.e., move these 
 https://github.com/pytorch/pytorch/blob/5f2fb58531c1ed472e979f82aa870f7cf50ce852/torch/utils/data/sampler.py#L114-L117

to

https://github.com/pytorch/pytorch/blob/5f2fb58531c1ed472e979f82aa870f7cf50ce852/torch/utils/data/sampler.py#L86-L91

lartpang replied:
Look like this:

```python
class RandomSampler(Sampler[int]):
    r"""Samples elements randomly. If without replacement, then sample from a shuffled dataset.
    If with replacement, then user can specify :attr:`num_samples` to draw.
    Args:
        data_source (Dataset): dataset to sample from
        replacement (bool): samples are drawn on-demand with replacement if ``True``, default=``False``
        num_samples (int): number of samples to draw, default=`len(dataset)`. This argument
            is supposed to be specified only when `replacement` is ``True``.
        generator (Generator): Generator used in sampling.
    """
    data_source: Sized
    replacement: bool

    def __init__(self, data_source: Sized, replacement: bool = False,
                 num_samples: Optional[int] = None, generator=None) -> None:
        self.data_source = data_source
        self.replacement = replacement
        self._num_samples = num_samples
        if generator is None:
            generator = torch.Generator()
            generator.manual_seed(int(torch.empty((), dtype=torch.int64).random_().item()))
        self.generator = generator

        if not isinstance(self.replacement, bool):
            raise TypeError("replacement should be a boolean value, but got "
                            "replacement={}".format(self.replacement))

        if self._num_samples is not None and not replacement:
            raise ValueError("With replacement=False, num_samples should not be specified, "
                             "since a random permute will be performed.")

        if not isinstance(self.num_samples, int) or self.num_samples <= 0:
            raise ValueError("num_samples should be a positive integer "
                             "value, but got num_samples={}".format(self.num_samples))

    @property
    def num_samples(self) -> int:
        # dataset size might change at runtime
        if self._num_samples is None:
            return len(self.data_source)
        return self._num_samples

    def __iter__(self):
        n = len(self.data_source)
        if self.replacement:
            for _ in range(self.num_samples // 32):
                yield from torch.randint(high=n, size=(32,), dtype=torch.int64, generator=self.generator).tolist()
            yield from torch.randint(high=n, size=(self.num_samples % 32,), dtype=torch.int64, generator=self.generator).tolist()
        else:
            yield from torch.randperm(n, generator=self.generator).tolist()

    def __len__(self):
        return self.num_samples
```

ejguan replied:
IMO, it's better to keep it in `__iter__` to make RNG lazy initialized.

lartpang replied:
In commit https://github.com/pytorch/pytorch/pull/53085/commits/60d010865f10897353aa7a2e8c1b5af073fc754c , I modified this part of the code.
But I realized that there was no need to construct a separate `genenrator` here, so I committed the changes again (https://github.com/pytorch/pytorch/pull/53085/commits/4645b7665861037676ceb372d201a9453ef0b4b7). 
I'm sorry that I didn't fill in the details in the commit massage here because I was in a hurry to commit.

lartpang committed:
Optimize the duplicate code.

Optimize the duplicate `generator` and `self.generator` in `_ iter__` of `RandomSampler`.

lartpang committed:
Update sampler.py

ejguan reviewed (approved):
Thank you for fixing the issue!

ejguan commented:
It breaks some internal tests, and I will look into it tmrw.

ejguan commented:
It seems that we have some tests relying on re-seeding of the RandomSampler for each iteration. So, we probably have to drop this patch and keep it as it is. Sorry about that.

lartpang commented:
> It seems that we have some tests relying on re-seeding of the RandomSampler for each iteration. 
I noticed that this commit https://github.com/pytorch/pytorch/pull/53085/commits/5f2fb58531c1ed472e979f82aa870f7cf50ce852 has passed the test, actually. The changes in it have prevented repeated initialization of the generator.

awaelchli commented:
Hi, what's the state of this PR? I'm struggling to understand how this change can break any existing logic. Assigning the generator to self is going to be very useful when we want to capture the state of a random sampler and then later resume from that state.

ejguan commented:
@awaelchli Thank you for bringing this PR back. My Github notification is overwhelmed and somehow I ignored @lartpang's comment. So sorry about that.
I have the same feeling that this one should not break any test.  @lartpang Could you please rebase this PR to master or viable/strict branch, then I can run the internal test. And I will investigate the broken tests.

lartpang commented:
@ejguan 

I have deleted the forked content, and re-operating git on the latest code base to modify the previous PR seems a bit complicated to me (sorry, I am not very familiar with git). 
I wonder if I can directly close the current PR and resubmit a new PR based on the latest version?

ejguan commented:
@lartpang 
You can close it and re-open a new one and request review to me.

If you have GitHub CLI installed. You can run `gh pr checkout https://github.com/pytorch/pytorch/pull/53085` to pull your PR back to local repo. Then, fetch the `origin viable/strict` and rebase onto it.

awaelchli commented:
@lartpang will we re-open a new one? I couldn't find it in the open PR section

lartpang commented:
@awaelchli 

I'm very sorry, but I have been caught up in other things recently.
Fortunately, after taking some time, I resubmitted it.

ejguan commented:
@lartpang @awaelchli
Finally the tests fixed and the PR is landed into master. It should be available in PyTorch nightly in the following days. Or, you can directly clone from master and compile PyTorch to enable this.
